### PR TITLE
Rename Dynamics to InPlaceDynamics

### PR DESCRIPTION
**Pull request by eb8680**
As part of changing the interface to `simulate` to be functional, this refactoring PR renames the abstract interface type `Dynamics` to `InPlaceDynamics` and separates out the `observation` method into a subtype. It also removes a type variable in the `Dynamics` type which was not being used.

### CODE REVIEW

File: chirho/dynamical/ops/dynamical.py [hunk 1]
```diff
@@ -107,18 +109,30 @@ def to_state(self) -> State[T]:
         return ret
 
 
+Dynamics = Callable[[State[S]], State[S]]
```
Comment (SamWitty) [inline chirho/dynamical/ops/dynamical.py:112]: Is this still used anywhere?

**Review by SamWitty — CHANGES_REQUESTED** [review]
Looks great. I like the distinction between `InPlaceDynamics` and `ObservableInPlaceDynamics` and the corresponding changes to the `StaticObservation` handler type.

Just one minor question about the use of the `Dynamics`.

File: chirho/dynamical/ops/dynamical.py (hunk 1 above)
Comment (eb8680) [reply chirho/dynamical/ops/dynamical.py]: I removed it

**Review by SamWitty — APPROVED** [review]
lgtm

**Pull request merged**